Handling of connection errors / Support reconnect better

obsoleted opened the pull request:
Starting as a fix for #38  but it may also relate to #39.

This PR makes it so the iOS plugin invokes the error callback for the connect exec when a read or write error is encountered during the message loop. It also makes it so that the plugin JS layer resets the connected state when the connect error callback is invoked. Together this allows an app to detect when the connection breaks and attempt to reconnect (by calling AllJoyn.connect again).

I pushed an update to the TV sample that does this here: https://github.com/obsoleted/cordova-tv-alljoyn/commit/489bffb569d843c87faf0f073dd29a38fa701ba0
# TODO
- [x] ~~Disconnect and invoke error when failing to send a message also~~ (Not really needed since there will be a failure callback as part of the regular flow. And the message loop will fail and disconnect as well)
- [ ] Check if Windows implementation needs this work and if so do it
- [ ] Check if Android implementation needs this work and if so do it
- [ ] Investigate and add test if possible
- [ ] Return consistent error messages when other apis are called but we are not connected

obsoleted committed:
Signal error when network issues detected and allow reconnect

- When the message loop encounters a network error (AJ_ERR_READ or
  AJ_ERR_WRITE) we invoke the error callback for the connect exec
- The connect method in AllJoyn.js now implements an error handler that
  resets the connected status to allow connecting again

vjrantal commented:
> when a read or write error is encountered during the message loop

To double-check, is the wanted behavior so that the `connect` error callback gets called if there is an error during the connection flow or at any time when the message loop runs?

Based on a quick look, it would be the former, but if the latter, I think the errors could be modeled using a "listener pattern" that we have for other types of events that may occur at any point of time and several times.

Would be nice to create automated tests for connection errors, but that requires a bit more work since we would need to signal from the "test execution context" to the host when the router should stop and when to start. This should be doable with an "HTTP bridge", but a bit of work.

Another (easier) option for testing is manual tests where the test flow "instructs" the user to manually start and stop router on the right phases of the test flow.

obsoleted commented:
This change has it so the `connect` error callback can be called either as a result of a bad `connect` or anytime a failure occurs in the msgloop. This error callback seemed to be the most reasonable because failures of this type generally put the app in the same state with respect to AllJoyn (i.e. back to where `connect` needs to be called again).

Having an error listener that we could invoke for any out of bound errors or messages would be reasonable also. Are you thinking it would be a specific "onConnectionLost" listener or something more generic? If more generic we would also have to start getting pretty well defined messages for the callback parameters.

I do like the listener based approach since it probably fits in better with how the callbacks are supposed to be invoked.

vjrantal commented:
> Are you thinking it would be a specific "onConnectionLost" listener

Yes. And to align with existing function names, maybe something like `bus.addConnectionErrorListener`.

vjrantal commented:
Continuing on above, I think it might be useful to separate an error during initial connection attempt (the error callback during connect-call that we have now) and an error that occurs later on at random point of time. An app might have different kind of error message in these two scenarios.

michiyosony commented:
We're trying to wrap our heads around what an `addConnectionErrorListener` would entail. 
Currently, this PR adds triggering the error callback on `AJ_ERR_READ` and `AJ_ERR_WRITE`.  
Are there other cases in which we know that the connection has been lost?  
For example, if the AJ router goes offline, an`AJ_ERR_READ` or `AJ_ERR_WRITE` isn't triggered until the bus is used (e.g. send a signal).
Would the proposed connection error listener be triggered when the router goes offline, or not until the bus is used?  More generally, at what point would it detect a connection error?

obsoleted commented:
The goal is for the listener to be invoked when the plugin detects a situation where connect needs to be called again. Indeed this PR starts with targeting the msg loop since that is where a great deal of errors will show up. There are unfortunately probably lots of edge cases for how an when errors show up. For example, with the next iteration of the PR i am working on (which I will update shortly) I find that when I kill the router the device is connected to it does trigger the connectionErrorListener without me explicitly making a new member invocation. This makes me wonder if there are other differences in play here (like router version).

I think with the way that the iOS plugin is currently structured there is always going to be a chance that the connection loss will be first realized during a member invocation or in some other non-message loop portion of the code. I think in these cases the message loop is likely to quickly encounter an error shortly after, but we could consider being more proactive when we find errors we _know_ mean connection loss.

obsoleted committed:
Moving connection error callback to a separate listener

- Based on feedback moving the connection error callback from the error
  callback of the connect function to a separate specific listener
  registration `addConnectionErrorListener`
- Also added checks in code for failing early if connection requiring
  methods are called when we are not connected

obsoleted commented:
I've pushed a new commit that has the listener based approach. Note that this does add a new exec that needs to be implemented in other platforms. So there is more work to be done for Windows and Android before merging it in (if even accepted).

michiyosony commented:
@obsoleted Interesting--what router version are you using?

obsoleted committed:
Merge branch 'master' into reconnect
